src/gurt/telemetry.c

```diff
@@ -682,6 +707,101 @@ d_tm_build_path(char **path, char *item, va_list args)
 	return rc;
 }
 
+/**
+ * Compute gauge statistics, min, max, mean, and standard deviation
+ * Uses B. P. Welford's algorithm for computing a running variance.
+ * This implementation computes an integer min and max with floating point
+ * mean and standard deviation.
+ *
+ * \param[in]	node		Pointer to a gauge node
+ */
+void
+d_tm_compute_gauge_stats(struct d_tm_node_t *node)
+{
+	uint64_t	value = 0;
+	double		sum_of_squares = 0.0;
+	double		variance = 0.0;
+	double		mean = 0.0;
+
+	if (node->dtn_metric->dtm_stats == NULL)
+		return;
+
+	value = node->dtn_metric->dtm_data.value;
+	node->dtn_metric->dtm_stats->sample_size++;
+
+	mean = node->dtn_metric->dtm_stats->mean +
+	       ((value - node->dtn_metric->dtm_stats->mean) /
+	       node->dtn_metric->dtm_stats->sample_size);
+
+	sum_of_squares = node->dtn_metric->dtm_stats->sum_of_squares +
+			 ((value - node->dtn_metric->dtm_stats->mean) *
+			 (value - mean));
+
+	node->dtn_metric->dtm_stats->mean = mean;
+	node->dtn_metric->dtm_stats->sum_of_squares = sum_of_squares;
+
+	if (node->dtn_metric->dtm_stats->sample_size > 2) {
+		variance = sum_of_squares /
+			   (node->dtn_metric->dtm_stats->sample_size - 1);
+		node->dtn_metric->dtm_stats->std_dev = sqrtl(variance);
+	}
```

Comment:
It sounds like a lot of calculation on the critical path. I checked the lustre code and they actually only maintain the sumsquare and probably delay the variance calculation to the post processing. Should we do that too and just collect the sum of square to avoid the impact?
Relevant links:
https://github.com/lustre/lustre-release/blob/master/lustre/include/lprocfs_status.h#L153
https://github.com/lustre/lustre-release/blob/master/lustre/obdclass/lprocfs_counters.c#L84

Reply:
That's a good idea.  I will go ahead and defer the variance and standard dev calculation until it is read.

Reply:
I read the Lustre code today.  I see that they leave the actual standard deviation calculation to some other software.  Interestingly, [https://wiki.lustre.org/Lustre_Monitoring_and_Statistics_Guide](url) says that 

> The standard deviation can be computed from sqrt({sum of values squared} - {mean value}²)

However, that formula is not correct.  It's more accurately: sqrt(({some of values squared} - (sample size * mean²) / (sample size -1)).  See if this new implementation is more to your liking.